### DISCUSSION TITLE
A discussion around if Cobra needs to be a hard dependency of Pocketbase

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#924] [Q&A]
After browsing through the start up code of a Pocketbase app into running a Cobra Command I noticed that Cobra really doesn't need to be baked in that closely.  I think that the two commands (server and migrate) could be broken out to be publicly callable.  Then all of that cobra stuff could be stuffed into say the `cmd` directory.  Then the cobra app setup could be called directly from there.

For myself, I'd prefer not to use cobra, and it would be just less work to get Pocketbase going if I didn't have to swap out cli frameworks.  It's only a method with say 4 variables that could be parameters, or a option struct if it allowed more long-term flexibility.

Maybe I'm missing a roadmap of why it's coupled so closely there, but I think that this could be better.  I'd rather think of Pocketbase as an Echo Server Plus than as it is now where it insists on hijacking the entire bootstrap.

In my particular case I need to move to Echo from Chi, which is unavoidable and I completely understand.  But the `urfave/cli` to `cobra` work is just going to be annoying for me and without any benefit (that I can see)

**Comment by ganigeorgiev** [comment]
We can wrap the api initialization in a helper function in the future but I don't think making cobra optional would be a good idea. It is too verbose and it will hurt the dev experience for those who just want to register a simple console command.

We'll need to have a default cli implementation somewhere no matter the library because you have to somehow at least take care for the base args parsing, base app initialization, graceful shutdown, etc. as in https://github.com/pocketbase/pocketbase/blob/master/pocketbase.go. Or in other words we'll always have at least one cli library as a dependency.

I'll add it in my todo to think on the proposal more thoroughly in the future, but for now it is a very low priority because it doesn't add too much to the project and there are way larger issues that need more attention (migrations flow improvement, filter rule enhancements, collection triggers/workflow builder, etc.).

Everyone is welcomed to develop a POC how this could look like, and although I may not merge the PR directly, we can always refer to it later or take some part of its implementation.

**Comment by dschissler** [comment]
On second thought, this seems more like a possible 2.0 refactor than something worth doing soon.  I think that the trick would be making it so that Cobra isn't on the main app struct.  That way you could still have the easy startup and then be able to offer a more hidden away startup for those who want to use it, and then Cobra doesn't make it into the binary because nothing references it.  Probably the Pocketbase config struct would need some kind of a New function to have the defaults be there instead of coming from the Cobra command.  At that point Pocketbase is much more "embedable" instead of being a full application container that always demands to be run.  I think that would allow developers who already have a big app and want to be able to embed PB such that it's easier to slowly migrate features from their old stack into the PB one from within the same program.

**Reply by IceWreck** [reply]
I think what you're describing can already be done 
https://github.com/pocketbase/pocketbase/discussions/1684#discussioncomment-4776798
